File: Έγγραφα SEM4105/Λύσεις ασκήσεων/06. Έκτο εργαστήριο/3/lapack.f90
PROGRAM eigenvalues
  IMPLICIT NONE

  INTERFACE
     SUBROUTINE DGEEV(JOBVL,JOBVR,N,A,LDA,WR,WI,VL,LDVL,VR,LDVR,&
          WORK, LWORK,INFO)
       IMPLICIT NONE
       CHARACTER,        INTENT (in)    :: JOBVL,JOBVR
       INTEGER,          INTENT (in)    :: N, LDA, LDVL, LDVR, LWORK
       INTEGER,          INTENT (out)   :: INFO       
       DOUBLE PRECISION, INTENT (inout) :: A(LDA,*)
       DOUBLE PRECISION, INTENT (out)   :: WR(*), WI(*)
       DOUBLE PRECISION, INTENT (out)   :: VL(LDVL,*), VR(LDVR,*), WORK(*)
     END SUBROUTINE DGEEV
  END INTERFACE

  integer, parameter :: n = 3
  integer, parameter :: ldvl = 1
  integer, parameter :: ldvr = 1

  DOUBLE PRECISION :: a(n,n), wr(n), wi(n), vl(ldvl,n), vr(ldvr,n), wrk(1)
  DOUBLE PRECISION, ALLOCATABLE :: work(:)
  INTEGER :: info,lwork
  
  INTEGER :: i
  
  a(1,1) = 6.3d0
  a(1,2) = 2.10d0
  a(1,3) = 4.15d0
  a(2,1) = 3.1d0
  a(2,2) = 5.14d0
  a(2,3) = 1.03d0
  a(3,1) = -11d0
  a(3,2) = 12.3d0
  a(3,3) = -8.8d0

  lwork = -1
  info = 0
  call dgeev('n', 'n', n, a, n, wr, wi, vl, ldvl, vr, ldvr, wrk, lwork, info)

  lwork = INT(wrk(1))
  ALLOCATE(work(lwork))
  info = 0
  call dgeev('n', 'n', n, a, n, wr, wi, vl, ldvl, vr, ldvr, work, lwork, info)
  DEALLOCATE (work)

  IF (info /= 0) THEN
     PRINT *, "error"
     STOP
  END IF
  
  DO i =1, N
     PRINT *, WR(i), WI(i)
  END DO
END PROGRAM eigenvalues

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/06. Έκτο εργαστήριο/3/lapack.c
#include <complex.h>
#include <stdio.h>
#include <stdlib.h>



void 
dgeev_(const char * jobvl, const char * jobvr, const int * n, 
       double a[], const int * lda, double wr[], double wi[], double vl[], 
       const int * ldvl, double vr[], const int * ldvr, double work[], 
       const int * lwork, int * info);


    
void eival(double a[], int n, double complex w[])
{
  double * wr = malloc(n*sizeof(double));
  double * wi = malloc(n*sizeof(double));
  
  int const ldvl = 1;
  double * vl = malloc(ldvl*n*sizeof(double));

  int const ldvr = 1;
  double * vr = malloc(ldvr*n*sizeof(double));

  double temp;
  int lwork = -1;
  int info = 0;
  char jobvl = 'n';
  char jobvr = 'n';
    
  dgeev_(&jobvl, &jobvr, &n, a, &n, wr, wi,
	 vl, &ldvl, vr, &ldvr, &temp, &lwork, &info);
    
  lwork = (int) temp;
  double * work = malloc(lwork*sizeof(double));

  info = 0;
    
  dgeev_(&jobvl, &jobvr, &n, a, &n, wr, wi,
	 vl, &ldvl, vr, &ldvr, work, &lwork, &info);

  if (info == 0) {
    for(int i=0; i < n; ++i) {
      w[i] = CMPLX(wr[i], wi[i]);
    }
  }

  free(work);
  free(vr);
  free(vl);
  free(wr);
  free(wi);
    
}


int main(void)
{
  size_t const n = 3;
  double * a = malloc(n*n*sizeof(double));

  a[0+3*0] = 6.3;
  a[0+3*1] = 2.1;
  a[0+3*2] = 4.15;

  a[1+3*0] = 3.1;
  a[1+3*1] = 5.14;
  a[1+3*2] = 1.03;

  a[2+3*0] = -11.0;
  a[2+3*1] = 12.3;
  a[2+3*2] = -8.8;


  double complex * w = malloc(n*sizeof(double complex));
    
  eival(a,n,w);

  for (size_t i = 0 ; i < n; ++i) {
    printf("%g %g\n", creal(w[i]), cimag(w[i]));
  }

  free(w);
  free(a);
    
  return 0;
}
